### PR TITLE
flake: allow specifying the lib argument to nixosSystem

### PR DESCRIPTION
**Pull request by r-vdp**
## Description of changes
Some configs rely on being able to pass their own lib argument into nixosSystem, for instance in order to add their own additional overlay to nixpkgs.lib.
This was broken by 039f73f134.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by roberth** [comment]
> Some configs rely on being able to pass their own lib argument into nixosSystem

They really shouldn't. What's the problem with those configs?

**Comment by r-vdp** [comment]
> > Some configs rely on being able to pass their own lib argument into nixosSystem
> 
> They really shouldn't. What's the problem with those configs? EDIT: I'd prefer to fix the root cause. Setting `lib` is prone to conflicts.

Why is that problematic? How would you suggest adding functions to lib then so that they are included in the lib that's passed as a module argument?
It seems to me that there should be an easy way to apply extra overlays to lib and use that modified lib everywhere.

**Comment by roberth** [comment]
> Why is that problematic?

Different "extensions" of `lib` may disagree about certain functions.
Furthermore, for any component to require a custom `lib` is bad UX, because it makes its problem "someone else's" problem.

> How would you suggest adding functions to lib

Make a PR to nixpkgs.

What is the use case?
Customizing `lib` (in an of itself) is probably not a life goal for you.

**Comment by r-vdp** [comment]
I don't really like this attitude. It's a lot of hassle to get things into nixpkgs, especially when they are not necessarily widely useful.

This attitude also makes it unnecessarily difficult to experiment with things or work on future contributions. I don't think it's your life goal either to protect people from themselves or from using nixpkgs in the way they like to (and maybe end up contributing back parts of it when they work out well).

So what's the downside of keeping things flexible, open, and extensible for those who want to experiment?
Do we really want to lock things down to the point where you need to fork the repo in order to add a function to lib?

**Comment by roberth** [comment]
> I don't really like this attitude.

I understand, and I would prefer not to deny functionality, but this is an important discussion to have.

> to experiment with things or work on future contributions.

For experimentation, could you use `specialArgs` to provide your own library separately? That provides a very sparsely populated namespace, that won't interfere with the actual `lib`.

When it's ready for contribution, you really should fork it, and write tests that call your new function(s). Other Nixpkgs contributors don't have access to your use case, and even if your calls are open source, you can't require them to test their improvements using those calls.

> I don't think it's your life goal either to protect people

I guess I had that coming with my choice of words :). It is not a life goal, but it is a sub-goal of mine to help users make future proof choices. That's why I work on documentation every now and then, and help review contributions to `lib` and to the entrypoints (such as this). I don't have a monopoly on the truth though, so that's why we're having a conversation.

**Comment by roberth** [comment]
> So what's the downside of keeping things flexible, open, and extensible for those who want to experiment?

I believe it already is even more flexible by different means, which cause less trouble for you and other users down the road.

> Do we really want to lock things down to the point where you need to fork the repo in order to add a function to lib?

If `lib` is just `lib`, everywhere, that's a lot easier to understand than everyone having their own customizations that may or may not be composable, and may or may not be propagated to other components, where they may or may not cause problems. It creates a lot of clarity.

File: flake.nix [hunk 1]
```diff
@@ -21,12 +21,13 @@
...
           import ./nixos/lib/eval-config.nix (
-            args // { inherit (self) lib; } // lib.optionalAttrs (! args?system) {
+            {
+              lib = final;
               # Allow system to be set modularly in nixpkgs.system.
               # We set it to null, to remove the "legacy" entrypoint's
               # non-hermetic default.
               system = null;
-            }
+            } // args
```
Comment (infinisil) [inline flake.nix:30]: Why would `lib = final` work? This would set it to the entire flakes output.

**Review by infinisil — COMMENTED** [review]
@roberth I think this is the wrong place to discuss this. Customising `lib` in this way has worked before and was broken with the linked commit. This is a regression (though it depends on experimental features) that should be fixed.

File: flake.nix (hunk 1 above)
Comment (r-vdp) [reply flake.nix:30]: `final` here is the final lib, it's the argument the `lib.extend` call, so it's essentially the same as `self.lib`.

Comment (infinisil) [reply flake.nix:30]: Ah right! I confused `self` with `final`

**Review by infinisil — APPROVED** [review]

**Comment by roberth** [comment]
Fair enough. It's a regression.
Could you add a test case somewhere?

**Comment by chapa** [comment]
I came across this issue too.

@roberth you are saying that configs shouldn't rely on the possibility of passing a `lib` argument to `nixosSystem`, which I tend to agree on the principle. How users can know whether they can or shouldn't rely on arguments of `nixpkgs.lib` functions (or even rely on functions themselves) ? Is there a documentation somewhere that I'm missing ?

**Comment by roberth** [comment]
It's more of a best practices thing to write code without undeclared dependencies, which is what a `lib` modification would tend to be.
You can assume that everything in the Nixpkgs `lib/` sticks around, but note that only the flake `lib` output contains anything NixOS related.

**Comment by r-vdp** [comment]
> Fair enough. It's a regression. Could you add a test case somewhere?

@roberth I wasn't exactly sure where to add the test, since the `nixosSystem` function is only defined in the flake's lib output.
I added it to the flake checks, you can run it with `nix build -vL '.#checks.x86_64-linux.nixosSystemCanPassLib'`.

Let me know if there's a better place to define this test.

File: flake.nix [hunk 2]
```diff
@@ -21,16 +21,35 @@
...
           );
       });
 
-      checks.x86_64-linux.tarball = jobs.tarball;
+      checks.x86_64-linux = {
+        tarball = jobs.tarball;
+        # Test that ensures that the nixosSystem function can accept a lib argument
```
Comment (roberth) [inline flake.nix:36]: This is a prominent location, so let's inform those who learn by example.

```suggestion
        # Test that ensures that the nixosSystem function can accept a lib argument
        # Note: prefer not to extend or modify `lib`, especially if you want to share reusable modules
        #       alternatives include: `import` a file, or put a custom library in an option or in `_module.args.<libname>`
```

File: flake.nix [hunk 3]
```diff
@@ -21,16 +21,35 @@
...
+            ifThisFunctionIsMissingTheTestFails = final.id;
+          });
+          modules = [
+            ./nixos/modules/profiles/minimal.nix
+            ({ lib, ... }: lib.ifThisFunctionIsMissingTheTestFails {
+              # Define a minimal config without eval warnings
+              nixpkgs.hostPlatform = "x86_64-linux";
+              boot.loader.grub.enable = false;
+              fileSystems."/".device = "nodev";
+              system.stateVersion = lib.versions.majorMinor lib.version;
```
Comment (roberth) [inline flake.nix:48]:
```suggestion
              # See https://search.nixos.org/options?show=system.stateVersion&query=stateversion
              system.stateVersion = lib.versions.majorMinor lib.version; # DON'T do this in real configs!
```

**Review by roberth — APPROVED** [review]

**Comment by roberth** [comment]
Thank you @r-vdp

**Pull request merged**